messages/off-chain-registry/message_manager.go

```diff
@@ -0,0 +1,199 @@
+)
+
+type messageManager struct {
+	logger             logging.Logger
+	destinationClients map[ids.ID]vms.DestinationClient
+	registryAddress    common.Address
+}
+
+func NewMessageManager(
+	logger logging.Logger,
+	messageProtocolConfig config.MessageProtocolConfig,
+	destinationClients map[ids.ID]vms.DestinationClient,
+) (*messageManager, error) {
+	// Marshal the map and unmarshal into the off-chain registry config
+	data, err := json.Marshal(messageProtocolConfig.Settings)
+	if err != nil {
+		logger.Error("Failed to marshal off-chain registry config")
+		return nil, err
+	}
+	var messageConfig Config
+	if err := json.Unmarshal(data, &messageConfig); err != nil {
+		logger.Error("Failed to unmarshal off-chain registry config")
+		return nil, err
+	}
+
+	if err := messageConfig.Validate(); err != nil {
+		logger.Error(
+			"Invalid off-chain registry config.",
+			zap.Error(err),
+		)
+		return nil, err
+	}
+	return &messageManager{
+		logger:             logger,
+		destinationClients: destinationClients,
+		registryAddress:    common.HexToAddress(messageConfig.TeleporterRegistryAddress),
+	}, nil
+}
+
+// ShouldSendMessage returns false if the version/protocol address tuple is already registered in the TeleporterRegistry contract.
+func (m *messageManager) ShouldSendMessage(unsignedMessage *warp.UnsignedMessage, destinationBlockchainID ids.ID) (bool, error) {
+	addressedPayload, err := warpPayload.ParseAddressedCall(unsignedMessage.Payload)
+	if err != nil {
+		m.logger.Error(
+			"Failed parsing addressed payload",
+			zap.Error(err),
+		)
+		return false, err
+	}
+	entry, destination, err := teleporterregistry.UnpackTeleporterRegistryWarpPayload(addressedPayload.Payload)
+	if err != nil {
+		m.logger.Error(
+			"Failed unpacking teleporter registry warp payload",
+			zap.Error(err),
+		)
+		return false, err
+	}
+	if destination != m.registryAddress {
+		m.logger.Info(
+			"Message is not intended for the configured registry",
+			zap.String("destination", destination.String()),
+			zap.String("configuredRegistry", m.registryAddress.String()),
+		)
+		return false, nil
+	}
+
+	// Get the correct destination client from the global map
+	destinationClient, ok := m.destinationClients[destinationBlockchainID]
+	if !ok {
+		return false, fmt.Errorf("relayer not configured to deliver to destination. destinationBlockchainID=%s", destinationBlockchainID.String())
+	}
+	client, ok := destinationClient.Client().(ethclient.Client)
+	if !ok {
+		panic(fmt.Sprintf("Destination client for chain %s is not an Ethereum client", destinationClient.DestinationBlockchainID().String()))
+	}
+
+	// Check if the version/protocol address tuple is already registered in the TeleporterRegistry contract.
+	registry, err := teleporterregistry.NewTeleporterRegistryCaller(m.registryAddress, client)
+	if err != nil {
+		m.logger.Error(
+			"Failed to create TeleporterRegistry caller",
+			zap.Error(err),
+		)
+		return false, err
```

Comment:
Would it make sense for all the errors in this function to wrap the error we return with the error message, and then log the error from the calling function?

Reply:
We don't really have a formal logging strategy in this repo, but in general we err on the side of over rather than under logging. What that translates to is that most of the data structures are composed of `Logger`s, and log internally before returning an error, which may lead to the same error being logged again.

As for wrapping, I think that would be a good idea, but I also think it would bloat the source code significantly when combined with the above logging strategy. We'd have to log the error wrapping text and then also wrap the error itself with the same text before returning.

Thinking about how to unify this, we could implement our own `Logger`, adding a new method `WrapError` that 1) logs the wrapped error and 2) returns the wrapped error so that it can easily be propagated. Something like:

```
func (l *LoggerWrapper) WrapError(wrapText string, err error) error {
    wrappedErr := fmt.Errorf("%s: %w", wrapText, err)
    l.Error(
        wrapText,
        zap.Error(err),
    )
    return wrappedErr
}
```

Regardless, I think the current implementation in this PR is generally consistent with how we're handling error logging in the rest of the repo, so I'd prefer to defer to a separate PR in which we unify our error logging across the entire repo at once.

Reply:
Yeah no problem for this PR. I've just been thinking about this a bit and would love a way to get the code size down a bit if possible.